sessiontxn/interface.go

```diff
@@ -15,44 +15,212 @@
+	txn, _ := p.Sctx.Txn(false)
+	return txn.Valid()
+}
+
+// OnStmtStart is the hook that should be called when a new statement started
+func (p *SimpleTxnContextProvider) OnStmtStart(ctx context.Context) error {
+	p.ctx = ctx
+	p.InfoSchema = p.Sctx.GetInfoSchema().(infoschema.InfoSchema)
+	p.stmtStarted = true
+	return nil
+}
+
+// OnStmtError is the hook that should be called when statement get an error
+func (p *SimpleTxnContextProvider) OnStmtError(_ error) {
+	return
+}
+
+// OnStmtRetry is the hook that should be called when a statement is retrying
+func (p *SimpleTxnContextProvider) OnStmtRetry(_ context.Context, _ error) error {
+	return nil
+}
+
+// Advise implements `TxnContextProvider.Advise`
+func (p *SimpleTxnContextProvider) Advise(opt AdviceOption, _ ...interface{}) error {
+	switch opt {
+	case AdviceWarmUpNow:
+		p.Sctx.PrepareTSFuture(p.ctx)
+	}
+	return nil
+}
+
+func UsingNonSimpleProvider(sctx sessionctx.Context) bool {
+	_, ok := GetTxnManager(sctx).GetContextProvider().(*SimpleTxnContextProvider)
+	return ok
+}
+
+type NewTxnRequest struct {
+	ExplictStart          bool
+	TxnMode               string
+	CausalConsistencyOnly bool
+	StaleReadTS           uint64
+	StaleReadInfoSchema   infoschema.InfoSchema
+}
+
 // TxnManager is an interface providing txn context management in session
 type TxnManager interface {
 	// GetTxnInfoSchema returns the information schema used by txn
 	GetTxnInfoSchema() infoschema.InfoSchema
+	// GetReadTS returns the read timestamp used by select statement (not for select ... for update)
+	GetReadTS() (uint64, error)
+	// GetForUpdateTS returns the read timestamp used by update/insert/delete or select ... for update
+	GetForUpdateTS() (uint64, error)
+	// IsTxnActive returns whether the current txn is active
+	IsTxnActive() bool
+	// ActiveTxn actives the txn
+	ActiveTxn(ctx context.Context) (kv.Transaction, error)
+
+	// EnterNewTxn enters a new txn
+	EnterNewTxn(ctx context.Context, request *NewTxnRequest) error
+	// OnStmtStart is the hook that should be called when a new statement started
+	OnStmtStart(ctx context.Context) error
+	// OnStmtRetry is the hook that should be called when a statement is retrying
+	OnStmtRetry(ctx context.Context, err error) error
+	// OnStmtError is the hook that should be called when statement get an error
+	OnStmtError(err error)
+	// Advise is used to provide some extra information to make a better performance.
+	// For example, we can give `AdviceWarmUpNow` to advice provider prefetch tso.
+	// Give or not give an advice should not affect the correctness.
+	Advise(opt AdviceOption, val ...interface{}) error
 
 	// GetContextProvider returns the current TxnContextProvider
 	GetContextProvider() TxnContextProvider
-	// SetContextProvider sets the context provider
-	SetContextProvider(provider TxnContextProvider) error
+	// ReplaceContextProvider replaces the context provider
+	ReplaceContextProvider(provider TxnContextProvider) error
```

Comment:
No it always returns nil. In the future we can forbid replacing to a new provider when one statement already started. But it is a little harder for current implement because `ReplaceContextProvider` is in `Compile` before `OnStmtStart`. They should be refactored later.